### ISSUE TITLE
Use AWSIotMqttManager.publishString with callback looks not work

### ISSUE THREAD

**Issue by pwa16000** [aws-amplify/aws-sdk-android#449]
### To help us solve your problem better, please answer the following list of questions.

* What service are you using?
Iot
* In what version of SDK are you facing the problem?
com.amazonaws:aws-android-sdk-iot:2.6.18

* Is the issue limited to Simulators / Actual Devices?
No

* Can your problem be resolved if you bump to a higher version of SDK?
No

* Is this problem related to specific Android/OS version?
No

* Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.
AWSIotMqttManager.publishString(java.lang.String str, java.lang.String topic, AWSIotMqttQos qos, AWSIotMqttMessageDeliveryCallback cb, java.lang.Object userData) looks not work.

If I use publishString(java.lang.String str, java.lang.String topic, AWSIotMqttQos qos), it works well, but if I use publishString with userData, after this line executed, AWS iot connection will be reset, then re-connection will be triggered.

Cloud watch log as below:
2018-04-18 09:04:34.911 TRACEID:6d0d4b78-849a-e61a-5d30-2e4609a1fb55 PRINCIPALID:1608872cd55dcf70ac96bdd510904c462fa228b6e4ba068720f72fb84ff6cb8d [ERROR] EVENT:PublishEvent TOPICNAME:test/responseSomething MESSAGE:PublishIn Status: CLIENT_ERROR
2018-04-18 09:04:34.911 TRACEID:6d0d4b78-849a-e61a-5d30-2e4609a1fb55 PRINCIPALID:1608872cd55dcf70ac96bdd510904c462fa228b6e4ba068720f72fb84ff6cb8d [INFO] EVENT:PublishEvent MESSAGE: IpAddress: 58.213.63.58 SourcePort: 46802
2018-04-18 09:04:34.919 TRACEID:5496fb81-a636-df2b-e4ea-6e0884e0cf1d PRINCIPALID:1608872cd55dcf70ac96bdd510904c462fa228b6e4ba068720f72fb84ff6cb8d [INFO] EVENT:MQTT Client Disconnect MESSAGE:Disconnect Status: SUCCESS
2018-04-18 09:04:34.919 TRACEID:5496fb81-a636-df2b-e4ea-6e0884e0cf1d PRINCIPALID:1608872cd55dcf70ac96bdd510904c462fa228b6e4ba068720f72fb84ff6cb8d [INFO] EVENT:MQTT Client Disconnect MESSAGE: IpAddress: 58.213.63.58 SourcePort: 46802

* Please include a stacktrace if applicable.

If you need help with understanding how to implement something in particular then we suggest that you first look into our [developer guide](https://docs.aws.amazon.com/mobile/sdkforandroid/developerguide/). You can also simplify your process of creating an application by using [Mobile Hub](https://console.aws.amazon.com/mobilehub/home#/).

**Comment by cbommas** [comment]
Hello @pwa16000 

Sorry to hear that you are having issues. Can you post a code snippet and the logcat from the device/Android Studio

**Comment by pwa16000** [comment]
Hello @cbommas 

Thanks very much for your analysis.

Attached is code snippet and log, if anything more is needed, please let me know.

Thanks
[info.txt](https://github.com/aws/aws-sdk-android/files/1936558/info.txt)

**Comment by pwa16000** [comment]
Hi, @kvasukib ，Thanks for your investigation. 

For your 1st question "Did you get any callback with failed delivery status when you called publishString with userData?", I did not get any callback with failed delivery status, or said, looks callback was not invoked. 

For the 2nd question, Network is fine, if I use AWSIotMqttManager.publishString(java.lang.String str, java.lang.String topic, AWSIotMqttQos qos) (without callback) instead of AWSIotMqttManager.publishString(java.lang.String str, java.lang.String topic, AWSIotMqttQos qos, AWSIotMqttMessageDeliveryCallback cb, java.lang.Object userData) , eveything is OK.

**Comment by kvasukib** [comment]
@pwa16000  Thank you for the reply.

1. "I did not get any callback with failed delivery status, or said, looks callback was not invoked." >>> We have found the bug in the code wherein an error condition we do not invoke the callback. I am working on to fix this issue.

2. publishString invokes publishData under the cover and both the sync and the async versions of the method calls the same function. But on error conditions, we have issues where we silently ignore instead of invoking the callback. This is related to (1). 

I will have an update once the fix is being made.
